### DISCUSSION TITLE
RuntimeError: main thread is not in main loop?

### DISCUSSION THREAD

**Discussion by zilch42** [zauberzeug/nicegui#2330] [Q&A]
### Question

Can anyone tell me what this error means?

I thought I knew at least where I introduced it but after restarting my session I'm seeing it more. If I understood it I might be able to get it into a small reproducible example.

```
Traceback (most recent call last):
  File "C:\Users\path\lib\tkinter\__init__.py", line 363, in __del__     
RuntimeError: main thread is not in main loop
Tcl_AsyncDelete: async handler deleted by the wrong thread
```

**Comment by falkoschindler** [comment]
How does this relate to NiceGUI? Without seeing any code leading to this RuntimeError in tkinter, it's hard to help.

**Reply by stepheweffie** [reply]
oh, hi

**Comment by stepheweffie** [comment]
Are you using Tkinter?

**Comment by zilch42** [comment]
> Are you using Tkinter?

Not explicitly, no. I wasn't sure what it was until I googled it. I assumed nicegui was, but there's no larger stack trace to confirm that.


> How does this relate to NiceGUI? Without seeing any code leading to this RuntimeError in tkinter, it's hard to help.

Fair question. Sorry, I was hoping someone might say "oh I've seen that before. Are you doing this perchance?".

My current theory is that it's because I'm doing extra post processing of data inside an `io_bound` call. When I load a large enough data set that the post processing takes > 1 sec then I start to see this error. However, I'm yet to be able to reproduce it in a simple app to share. 

I think the post processing should ideally go in a `cpu_bound` call, but I've been struggling to get that to work. I've just posted another question about that  #2358 so if I can get that to work then this might go away, or become clearer. I'll add to this if I ever get something reproducible.

**Comment by jcderner** [comment]
I have the same issue. I guess (but could not yet verify) it comes from using matplotlib or pyplot which uses as default TKAGG as backend. (S. also e.g. https://stackoverflow.com/questions/52839758/matplotlib-and-runtimeerror-main-thread-is-not-in-main-loop )
Maybe niceGUI can make the backend somehow configurable?

**Reply by jcderner** [reply]
Putting a proper matplotlibrc file with "backend: Agg" seems to help.
